networks/3a_validate_array_vs_rnaseq_modules.py

```diff
@@ -0,0 +1,217 @@
+    y=1.05,
+)"""
+
+# Plot heatmap
+"""h1b = sns.clustermap(pao1_rnaseq_corr_log_spell.abs(), cmap="viridis", figsize=(20, 20))
+h1b.fig.suptitle(
+    f"log transform + SPELL corrected using {num_singular_values_log} vectors (PAO1 rnaseq)",
+    y=1.05,
+)"""
+
+# ## Clustering and get module membership
+
+# Clustering using DBSCAN
+if cluster_method == "dbscan":
+    pao1_array_clustering = DBSCAN(eps=density_threshold).fit(pao1_array_corr_log_spell)
+    pao1_rnaseq_clustering = DBSCAN(eps=density_threshold).fit(
+        pao1_rnaseq_corr_log_spell
+    )
+
+# Clustering using hierarchal clustering
+if cluster_method == "hierarchal":
+    pao1_array_clustering = AgglomerativeClustering(
+        n_clusters=None, distance_threshold=hier_threshold, linkage=link_dist
+    ).fit(ppao1_array_corr_log_spell)
+    pao1_rnaseq_clustering = AgglomerativeClustering(
+        n_clusters=None, distance_threshold=hier_threshold, linkage=link_dist
+    ).fit(pao1_rnaseq_corr_log_spell)
+
+# Clustering using affinity propogation
+if cluster_method == "affinity":
+    pao1_array_clustering = AffinityPropagation(random_state=0).fit(
+        pao1_array_corr_log_spell
+    )
+    pao1_rnaseq_clustering = AffinityPropagation(
+        random_state=1, damping=affinity_damping
+    ).fit(pao1_rnaseq_corr_log_spell)
+
+# +
+# Get module membership for a single threshold
+# Format and save output to have columns: gene_id | group_id
+pao1_array_membership_df = pd.DataFrame(
+    data={"module id": pao1_array_clustering.labels_},
+    index=pao1_array_corr_log_spell.index,
+)
+
+pao1_array_membership_df["module id"].value_counts()
+# -
+
+pao1_array_membership_df.head()
+
+# +
+pao1_rnaseq_membership_df = pd.DataFrame(
+    data={"module id": pao1_rnaseq_clustering.labels_},
+    index=pao1_rnaseq_corr_log_spell.index,
+)
+
+pao1_rnaseq_membership_df["module id"].value_counts()
+# -
+
+pao1_rnaseq_membership_df.head()
+
+# ## Compare composition of modules
+#
+# For a given array module, are the genes within 1 module in the RNA-seq compendium?
+
+# +
+array_module_mapping = {}
+for grp_id, grp in pao1_array_membership_df.groupby("module id"):
+    grp_mapped = pao1_rnaseq_membership_df.loc[grp.index]
+    array_module_mapping[grp_id] = list(grp_mapped["module id"].unique())
+
+array_module_mapping
+
+# +
+# Are there any array modules that map to a single rnaseq module?
+consistent_array_modules = []
+for array_module_id, list_rnaseq_module_ids in array_module_mapping.items():
+    if len(list_rnaseq_module_ids) == 1:
+        print(array_module_id, list_rnaseq_module_ids)
+        consistent_array_modules.append(array_module_id)
+
+print(len(consistent_array_modules) / len(array_module_mapping))
+
+# +
+rnaseq_module_mapping = {}
+for grp_id, grp in pao1_rnaseq_membership_df.groupby("module id"):
+    grp_mapped = pao1_array_membership_df.loc[grp.index]
+    rnaseq_module_mapping[grp_id] = list(grp_mapped["module id"].unique())
+
+rnaseq_module_mapping
```

Comment:
I'm reading aloud this piece of code to see if I understand: for rna-seq data, `grp_id` is the cluster id, `grp` contains the list of genes in this cluster, and when you do `grp.index` you are getting that list of genes in the cluster. So `rnaseq_module_mapping[grp_id]` contains the list of cluster ids in array data that contain each of these genes in the cluster `grp_id` in rnaq-seq data.

Reply:
So what you would like to find is that each single rna-seq cluster map to a single array cluster, right? This is what you call "consistency"?

Reply:
Yes your interpretation is correct. I'll try to add more clarify to this.